src/main/java/games/strategy/triplea/oddsCalculator/ta/ConcurrentOddsCalculator.java

```diff
@@ -1,437 +1,49 @@
-  // do not let multiple calculations or setting calc data happen at same time
-  private final Object mutexCalcIsRunning = new Object();
-  private final List<OddsCalculatorListener> listeners = new ArrayList<>();
-
-  public ConcurrentOddsCalculator(final String threadNamePrefix) {
-    executor = Executors.newFixedThreadPool(MAX_THREADS,
-        new DaemonThreadFactory(true, threadNamePrefix + " ConcurrentOddsCalculator Worker"));
-    logger.fine("Initialized executor thread pool with size: " + MAX_THREADS);
-  }
-
-  @Override
-  public void setGameData(final GameData data) {
-    // increment so that a new calc doesn't take place (since they all wait on this latch)
-    latchSetData.increment();
-    // cancel any current setting of data
-    --cancelCurrentOperation;
-    // cancel any existing calcing (it won't stop immediately, just quicker)
-    cancel();
-    synchronized (mutexSetGameData) {
-      try {
-        // since setting data takes place on a different thread, this is our token. wait on it since
-        latchWorkerThreadsCreation.await();
-        // we could have exited the synchronized block already.
-      } catch (final InterruptedException e) {
-        Thread.currentThread().interrupt();
-      }
-      cancel();
-      isDataSet = false;
-      isCalcSet = false;
-      if (data == null || isShutDown) {
-        workers.clear();
-        ++cancelCurrentOperation;
-        // allow calcing and other stuff to go ahead
-        latchSetData.countDown();
-      } else {
-        ++cancelCurrentOperation;
-        // increment our token, so that we can set the data in a different thread and return from this one
-        latchWorkerThreadsCreation.increment();
-        executor.submit(() -> createWorkers(data));
-      }
-    }
-  }
+  private boolean cancelled = false;
 
+  /**
+   * Concurrently calculates odds using the OddsCalculatorWorker. It uses Executor to process the results. Then waits
+   * for all the future results and combines them together.
+   */
   @Override
-  public int getThreadCount() {
-    return currentThreads;
-  }
+  public AggregateResults calculate(final OddsCalculatorParameters parameters) {
+    final ExecutorService executor = Executors.newFixedThreadPool(MAX_THREADS);
+    final long start = System.currentTimeMillis();
+    final AggregateResults aggregateResults = new AggregateResults(parameters.runCount);
 
-  // use both time and memory left to determine how many copies to make
-  private static int getThreadsToUse(final long timeToCopyInMillis, final long memoryUsedBeforeCopy) {
-    if (timeToCopyInMillis > 20000 || MAX_THREADS == 1) {
-      // just use 1 thread if we took more than 20 seconds to copy
-      return 1;
+    final int totalRunCount = parameters.runCount;
+    for (int i = 0; i < totalRunCount && !cancelled; i++) {
+      parameters.setRunCount(1);
+      executor.submit(() -> aggregateResults.addResult(new OddsCalculator().doSimulation(parameters)));
```

Comment:
Since `ExecutorService#submit` is not blocking, the cancelled boolean won't do much.
Solution to this would be to either make the Executor service a field, which is shared among all calls to calculate and to call `ExecutorService#shutdownNow` in the cancel method, or to create a Collection (LinkedList or Set, not sure which is better), which temporarily stores all instances of OddsCalculator and calls cancel on them (probably very resource intensive)

Reply:
Good catch, cancel was broken. dc6a183 fixes it.